src/irmin-pack/inode.ml

```diff
@@ -431,6 +431,50 @@ struct
 
     let stable t = t.stable
 
+    module Tree = struct
+      type kind = Contents of metadata | Node [@@deriving irmin]
+      type entry = { name : step; kind : kind; hash : hash } [@@deriving irmin]
+
+      type 'a pointer = { index : int; pointer : hash; tree : 'a }
+      [@@deriving irmin]
+
+      type 'a tree = { depth : int; length : int; pointers : 'a pointer list }
+      [@@deriving irmin]
+
+      type t = Tree of t tree | Value of entry list [@@deriving irmin]
+
+      let entry (name, v) =
+        match v with
+        | `Contents (hash, m) -> { name; kind = Contents m; hash }
+        | `Node hash -> { name; kind = Node; hash }
+    end
+
+    let to_tree (la : 'ptr layout) (t : 'ptr t) =
+      let rec aux t =
+        match t.v with
+        | Tree tr ->
+            ( Lazy.force t.hash,
+              Tree.Tree
+                {
+                  depth = tr.depth;
+                  length = tr.length;
```

Comment:
Looks good to me but do we need these two fields that can be easily retrieved?

Reply:
I think it's useful to match the spec as closely at possible here.